components/dsp_processor/dsp_processor.c

```diff
@@ -236,17 +253,42 @@ int dsp_processor_worker(void *p_pcmChnk, const void *p_scSet) {
   // volatile needed to ensure 32 bit access
   volatile uint32_t *audio_tmp =
       (volatile uint32_t *)(pcmChnk->fragment->payload);
-  dspFlows_t dspFlow;
+  
+  // Local working copy of filter parameters
+  static filterParams_t currentFilterParams = {0};
+  static bool paramsInitialized = false;
+  
+  // Initialize on first run
+  if (!paramsInitialized) {
+    currentFilterParams.dspFlow = all_params.active_flow;
+    currentFilterParams.fc_1 = all_params.flow_params[all_params.active_flow].fc_1;
+    currentFilterParams.gain_1 = all_params.flow_params[all_params.active_flow].gain_1;
+    currentFilterParams.fc_2 = all_params.flow_params[all_params.active_flow].fc_2;
+    currentFilterParams.gain_2 = all_params.flow_params[all_params.active_flow].gain_2;
+    currentFilterParams.fc_3 = all_params.flow_params[all_params.active_flow].fc_3;
+    currentFilterParams.gain_3 = all_params.flow_params[all_params.active_flow].gain_3;
+    paramsInitialized = true;
+  }
 
-  // check if we need to update filters
-  if (xQueueReceive(filterUpdateQHdl, &filterParams, pdMS_TO_TICKS(0)) ==
-      pdTRUE) {
+  // If parameters were changed by API, copy them from centralized storage
+  if (params_changed) {
```

Comment:
I think you need to guard this by a mutex.
Or maybe better use a mutex for signaling:
`if xSemaphoreTake(update_params_mutex, pdMS_TO_TICKS(0) {...}`
and in `init` you need to initialize and take the mutex and in `dsp_processor_update_filter_params` you give the mutex.

Reply:
binary semaphore might be better than mutex

Reply:
Changed, please have a look. Not sure I'm skilled enough to see the benefit of the binary semaphore, did the way I can ;)

Reply:
That was not what I meant. Please see my suggestions:

```suggestion
  if (xSemaphoreTake(paramsChangedSemaphoreHandle, 0) == pdTRUE) {
```